Project : RecipeNation (External Code Review)

AmineHLub opened the pull request:
<h1> Summary </h1>

- As a team we setup the linters / webpack configurations / Jest.
- As a team we made header and footer (shared code).

@**mahbubul14** Created : 
- Getting & Displaying random meals from API
- based on each **MealAPI's ID**, user gets likes given to that **particular ID** stored in **Microverse's Intervention's API.**
- On click on 'heart logo' (Like) Changes to the number of likes happen to the **Intervention's API.**
- Function to calculate changes (tested)
- Post the changes immediately to the DOM interface.

@AmineHLub Created :
- On click on on either **MEAL image** or **comment button**, a **pop modal shows up** containing :
- Meal **information** (from **MEAL API**)
- A **commentCalculator function** will **calculate** how many comments there are for that particular meal (**based on its ID**) (**tested**) .
- The function would then **post** the **number of the comments and the comments themselves** back to the comments section at the popup.
- Eventually an **Add-comment function**.

**PS** : The API servers are a bit slow which might in **rare cases introduce bugs.** please do consider that. Consequential POST and GET request for updating ran asynchronously does not help as even if the POST is processed from the client side. **_At the server side it might need additional time to be processed_**. Therefore we used setTimeout function giving it half a second time to have the sequence fully process at both sides (Client - Server)

<h1> Video </h1>

- **link for video presentation** (**UPDATED** : Trimmed and Added part to speak about some code's features ) : https://drive.google.com/file/d/1Oh6MHEAeAGA13YPCUx_E9urFfR9CypRC/view?usp=sharing
Also on **Youtube** : https://www.youtube.com/watch?v=1yGasnNwUHo (**better 720p quality**)

AmineHLub committed:
Setup webpack, linters, jest for testing

AmineHLub committed:
Create logo using Adobe sparks (include license file)

mahbubul14 committed:
Add header

AmineHLub committed:
Add footer, change header colors, add boxshadow, fix some padding bugs

AmineHLub committed:
Fix linter error, build project at /dist

AmineHLub committed:
Merge pull request #22 from AmineHLub/footer-header

Footer-header

AmineHLub committed:
Add static modal creating function on click on 'Meals', close button works too

mahbubul14 committed:
Add a js file

AmineHLub committed:
Adding data from API (static-link)

AmineHLub committed:
Build at /dsit

AmineHLub committed:
Async function change

AmineHLub committed:
Revert unecessary changes to fix linters errors as it was a bug

mahbubul14 committed:
List update from api

mahbubul14 committed:
Update display list

mahbubul14 committed:
Update loop and add random api

mahbubul14 committed:
Fix linters error

mahbubul14 committed:
Update linters yml file

mahbubul14 committed:
Update linters file

mahbubul14 committed:
Revert "Fix linters error"

This reverts commit ccd50158f1b9c3605109d99479a05f8312e5a47d.

AmineHLub committed:
fixing linters attempt#1

AmineHLub committed:
fixing linters #2

AmineHLub committed:
packages-reload

mahbubul14 committed:
Fix all linters error and  update json files

mahbubul14 committed:
Merge pull request #25 from AmineHLub/d-list

Display list of items on the Homepage - Student A

AmineHLub committed:
Getting changes from dev branch

AmineHLub committed:
Fix items positioning at index.js

AmineHLub committed:
Re-direct the click event from meal to the appropriate click for each food item, increase the number from 6 to 9, fix stlying and add hovering effects, better popup UI

AmineHLub committed:
Add sticky navbar

AmineHLub committed:
Add cursor pointer whenever the user hover an image

AmineHLub committed:
Merge pull request #23 from AmineHLub/popup

Popup 1

AmineHLub committed:
Add messages Statically and style them

AmineHLub committed:
Add comments dynamically from template pre-POSTed API id (not a real mealId)

AmineHLub committed:
Merge pull request #26 from AmineHLub/comments-on-popup

Comments on popup

AmineHLub committed:
Enhance comments style

AmineHLub committed:
Merge pull request #27 from AmineHLub/comments-on-popup

Enhance comments style

mahbubul14 committed:
Add likes to DOM using pre-Post on likes on intervention API

mahbubul14 committed:
Merge pull request #28 from AmineHLub/like-count

[4pt] - Display number of likes for each item on the Homepage

mahbubul14 committed:
Add like button

AmineHLub committed:
Add full Adding functionality, Add comments counter function and export it

AmineHLub committed:
Add tests

Update commentsCounter.js function : Add try and catch

Add fetch-mock to package.json

AmineHLub committed:
Update documentation to require installing jest-fetch-mock for testing

AmineHLub committed:
Replace button with anchor link 'a' for performance gain

mahbubul14 committed:
Add likes interaction

AmineHLub committed:
Merge pull request #29 from AmineHLub/add-comments

Comments Adding :100:

AmineHLub committed:
Merge pull request #31 from AmineHLub/addLike

Create feature: add new like - Student A

AmineHLub committed:
Replace form with normal div as form causes unexpected reloading issues

AmineHLub committed:
Merge pull request #32 from AmineHLub/add-comments

BUG : Fixed

AmineHLub committed:
Fix issue when you like an item from the bottom of the page it takes you back to the upper side

AmineHLub committed:
Add fontAwesome Attribution in documentation

mahbubul14 committed:
Add rules for testing

AmineHLub committed:
Delete getLikesMock.test.js

AmineHLub committed:
Delete getLikesMock.js

mahbubul14 committed:
Add testing

AmineHLub committed:
Merge pull request #33 from AmineHLub/test-item

Add tests for items counter - Student A

AmineHLub committed:
Deploy Project and update Demo link

Minor comment btn fix

AmineHLub committed:
Fix conflict

AmineHLub committed:
Add project Pictures

AmineHLub committed:
Update Documentation

AmineHLub committed:
Delete consolelogs

helman101 commented on dist/index.html:
```diff
@@ -0,0 +1,31 @@
+        <nav class ='nav-link'>
+            <ul class ="d-flex">
+                <li><a href="#"><img src="assets/main-logo2a04e92b103dd4a80351.png" alt="logo"></a></li>
+                <li><a href="#">Meals</a></li>
```
- [ ] As you did with comments, you should have a counter for the number of meals on the page, kindly fix this issue. 😄 
    - Remember that even if the API gives you the number of items, you should create a specific function for that purpose.
    - And don't forget that you should test this function. 🚀 
    ![image](https://user-images.githubusercontent.com/60789317/137360083-b90e4c6c-ac2b-4953-9aa5-ef798750c72e.png)

AmineHLub replied:
Excuse me the API does not give you any set number of items so we instead used the get random meal breakpoint and manually set it to 9

https://www.themealdb.com/api.php

it only does : 
"Lookup a selection of 10 random meals (only available to $2+ Patreon supporters) give you 10 random meals". Which is not free!. Besides we instead did the function count for the likes which the API does give and calculated it.

helman101 replied:
@AmineHLub Well, if that is the case, please add that number to the navbar, like is showed in the image, and don't worry about the counter function, just make it similar to the template. 😄

AmineHLub replied:
@helman101 Thank you very much for your kind review bro I will do it this way

helman101 replied:
@AmineHLub You're welcome, let me know if you need anything else, I will be happy to give you a hand ✋

helman101 commented on src/__mocks__/commentsCounterMock.js:
```diff
@@ -0,0 +1,14 @@
+  }
+};
+
+module.exports = commentsCounterMock;
```
- [ ] Please remember to use ES6 syntax, ES6 syntax makes your code more readable along with other advantages. 😸

helman101 commented on src/__mocks__/getLikesMock.js:
```diff
@@ -0,0 +1,16 @@
+    // return likes;
+  }
+};
+module.exports = getLikesMock;
```
- [ ] Please remember to use ES6 syntax, ES6 syntax makes your code more readable along with other advantages. 😸 
    - Please take a look to the rest of your project and change this syntax for ES6 syntax 🚀

AmineHLub replied:
I tried to do so. however it kept denying it as I would also need to transpile it using bable I suppose. Using ES6 is not an absolute requirements. it's an objective which we have achieved at least for 90% of the project : 
Arrow functions, importing, modules,object destructuring... I will definitely try it again though. Thanks for the feedback

helman101 replied:
@AmineHLub Using ES6 syntax is a requirement, if you need to transpile these files I recommend you to take a look at this documentation [Using ES6 with jest](https://jestjs.io/docs/getting-started#using-babel) 😸

helman101 reviewed (changes requested):
# STATUS: CHANGES REQUIRED ♻️ 

Hi @AmineHLub and @mahbubul14   👋, 

Nice work so far on this milestone of the project  😄, but there are still some changes to do.

### Required Changes ♻️

- [ ] You have a nice explanation of your project features, but there are 2 things i like to highlight about your video.
    - First issue, the duration of your video is more than 5 minutes, remember that as the project requirements said, the max length is 5 min.
    
        - ![image](https://user-images.githubusercontent.com/60789317/137356598-c485a189-a000-4794-b61c-abb5ac47cb13.png)
    - You have done an excellent job showing and explaining your deployed project, but you forgot to highlight and explain a piece of code that you feel proud of.
    
        - ![image](https://user-images.githubusercontent.com/60789317/137356979-6e9554e4-4b92-4695-9053-d59e0588aa83.png)

### Optional Changes 🤓 

- [ ] [OPTIONAL] The button that closes the popups menus does not have any interactions, I would recommend adding `cursor: pointer` attribute to the class and also change the color in `:hover` so it will see like something you can interact with. 



Cheers and Happy coding!👏👏👏

![gif](https://media.giphy.com/media/13HgwGsXF0aiGY/giphy.gif)

If anything is not 100% clear please leave your questions or comments in the PR thread, we will be glad to help you 😄.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

AmineHLub committed:
Add effects to close icon button

AmineHLub committed:
Add bable to transpile ES6

AmineHLub committed:
Update documentation to require Bable (for testing purposes)

AmineHLub committed:
Fix webpack config file

carlos-ssh commented on src/addNewComment.js:
```diff
@@ -0,0 +1,22 @@
+  const commentInput = document.querySelector('.adding-comment').value;
+  const newmsg = new MessageObject(id, usernameInput, commentInput);
+
+  fetch('https://us-central1-involvement-api.cloudfunctions.net/capstoneApi/apps/5OvdDl22dcA0TUpGjcko/comments', {
```
- Comments aren't loading kindly work on it:

![Screen Shot 2021-10-14 at 20 26 29](https://user-images.githubusercontent.com/22258274/137417717-7a987987-5e3c-4a6f-89bb-1ccce3cf5c83.png)

AmineHLub replied:
@carlos-ssh  Kindly, can you explain to me what do you mean by comments aren't showing up ? Did you add a comment and it did not show up ? or in which position ?

AmineHLub replied:
Hey @carlos-ssh
The Error Provided there is 400. Fetch detects only network errors. Other errors (401, 400, 500) should be manually caught and rejected. and it's not because the project does not show the comment it's just that that particular meal does not have any and if you fetch it will return this json object : 
```
{
"error": {
"status": 400,
"message": "'item_id' not found."
}
}
```
example of meal that already has comments stored to the api : 
![image](https://user-images.githubusercontent.com/83329154/137477091-c0bb15d2-25a2-4951-b0ed-5e0e26b49fb2.png)

**_Comments are shown and no error messages_**

different example of a meal that did not happen to have a previous comment, error 400 is shown and no comments of course : 
![image](https://user-images.githubusercontent.com/83329154/137477298-4ac90cfe-348e-4e98-8f26-70cbf9f968f4.png)

carlos-ssh reviewed (changes requested):
Hello team, 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.


 ## Required Changes ♻️
_Check the comments under the review._


![changesrequested](https://user-images.githubusercontent.com/22258274/137417979-f3f6a402-ea55-4874-95cc-a2b020b28b45.gif)




## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have only 3 code reviews per project. If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

AmineHLub committed:
Add some modification to seek error and catch it at commentCounter.js

This would not solve getting 400 as it's the server's response of fetching an element that does not have comments but would make the code not wait for a json file of an error request

AmineHLub committed:
Remove if statement that would check if the returned value of the fetching function is a number as (catch) would do send back an empty array

wandji20 reviewed (approved):
## Hi @AmineHLub and @mahbubul14  :raised_hands:

##### You have done exceptionally well, and have met all requirements for this project :+1:

# Status: Approved 👍 :partying_face: 

Your project has been approved, however, Keep up the good work.

## Congratulations :+1:  :fireworks: 

_Go ahead and merge this branch. Also, don't forget to submit the project completion form.
Good luck with your next project :mechanical_arm:  :slightly_smiling_face:_ 

 If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
